### PR TITLE
Qr code reader

### LINKED ISSUE #408: Possibility to use camera as QR code scanner instead of taking photos
In some cases, users want to scan a QR code and parse the content as a string to the form. It will be good to have an option to use QR scanner feature from the camera, in addition to photo taking.

Not sure, if this will need extending QGIS widget types??

### PR DESCRIPTION
**Pull request by sklencar**
QR code reader implementation using ZXing library.
* Supported multiple platforms
    * full support for Android and iOS,
    * supported also MacOS desktop build, but seems to have bug that it is capturing invalid image for ZXing library,
    * WIN build not supported;
* available for TextEdit widget if an attribute's `name` contains `qrcode` case insensitive. Note that `name` is taken from field.name or overwriten by field.alias if exists.
* added invalid data changed notification
* basic style with simple header and camera overlay

Related manual-test: https://github.com/lutraconsulting/input-manual-tests/issues/28
closes #408 

Known issues
* camera issues
    * blank page when camera is being loaded (affects some devices)
    * `QImage qt_imageFromVideoFrame(const QVideoFrame &) : unsupported pixel format Format_ABGR32` while converting videoOutput to QImage (seems to be Android only).
    * sometimes not QR decoder not responses well

https://user-images.githubusercontent.com/6735606/108194128-8769a900-7116-11eb-8151-dc3987f31e13.MP4

### CODE REVIEW

**Review by tomasMizera — COMMENTED** [review]
Very good job!

I did not go through the whole code yet, though what I have seen so far looks great. I tested it on Android and it works well. I could not get Linux build running but I guess we will be able to sort it out.

Some notes:
 - Permission for camera on Android is now again requested at startup (this was changed in #1081). I think what we want is to opt it when needed - when user clicks on `take a photo` or now qr code icon. It is possible that it comes from ZXing library, but the problem is that it will be forbidden in upcoming version of Android to immediately ask for permissions without any information. 
 - If permission is not granted and user opens QR code scanner, it says something like "Camera service is not operating?" which is not true, user only did not provide permission.
 - Maybe icon for QR Code can be smaller?
 - When user scans invalid (not convertible) value for field, maybe we could display some more user friendly message, now:

File: .github/workflows/android.yml [hunk 1]
```diff
@@ -1,7 +1,7 @@
 name: Android
 on: [push]
 env:
-  SDK_VERSION: android-20
+  SDK_VERSION: android-21
```
[the remark this answers is not in the document]
Comment (sklencar) [reply .github/workflows/android.yml:4]: SDK has to be upgraded due to xzing lib

**Comment by sklencar** [comment]
All remarks from notes should be resolved. However I kept the QR code icon unchanged. I would suggest to revise all icons in quick widgets to use its size variable or derive size from a ratio of rowHeight.

**Review by tomasMizera — COMMENTED** [review]
Based on a phone discussion, please add some more comment to have an idea how this is working. It is working very nice on Android! Thanks for resolving the permission problems, it is now working as expected.

**Comment by tomasMizera** [comment]
Also regarding the manual test, I vote to add it here https://github.com/lutraconsulting/input-manual-tests/issues/5 rather than creating new test case. Thanks!

**Comment by sklencar** [comment]
Remarks resolved + manual test relocated.

I have checked the issue with scanning the same code for multiple fields. The issue seems to be rather linked with the library (or, of course, something else). This use case is possible, but the scanner acts like it has problem to scan it sometimes. Usually when I was moving camera a bit and wait longer time, it eventually scan the code. This could be possibly improved to add some flags to the DecodeHints in decoder (e.g Binarizer which I keep as default since I was having some issues with reading barcodes when i set it to Binarizer::FixedThreshold). In any case, this is something that has to be investigated further.

File: app/codefilter.cpp [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    }
+
+    QVideoFrame run( QVideoFrame *input,
+                     const QVideoSurfaceFormat &surfaceFormat,
+                     QVideoFilterRunnable::RunFlags flags ) override
+    {
+      Q_UNUSED( surfaceFormat );
+      Q_UNUSED( flags );
+
+      if ( mFilter->decoder()->isDecoding() )
```
Comment (tomasMizera) [inline app/codefilter.cpp:33]: Check if `mFilter` is a valid pointer

File: app/codefilter.cpp [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+CodeFilter::CodeFilter()
+{
+  mDecoder = new QRDecoder;
+
+  QObject::connect( mDecoder, &QRDecoder::capturedChanged, this, &CodeFilter::setCapturedData );
+}
+
+QVideoFilterRunnable *CodeFilter::createFilterRunnable()
+{
+  return new QRRunnable( this );
```
Comment (tomasMizera) [inline app/codefilter.cpp:64]: We should probably save this address to a new attribute and free it in `CodeFilter`'s destructor or somehow destroy QRRunnable instance when `this` (CodeFilter) is deleted

File: app/codefilter.cpp [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+#include <QtMultimedia/qvideoframe.h>
+#include <QImage>
+#include <QOpenGLContext>
+#include <QOpenGLFunctions>
+#include <QDebug>
+
+void processImage( QRDecoder *decoder, const QImage &image )
+{
+  decoder->process( image );
```
Comment (tomasMizera) [inline app/codefilter.cpp:13]: Check if `decoder` is a valid pointer

File: app/codefilter.h [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+  signals:
+    void capturedDataChanged();
+    void isDecodingChanged( bool isDecoding );
+
+  private slots:
+    void setCapturedData( const QString &capturedData );
+
+  private:
+    QSharedPointer<QMutex> _mutexP;
```
Comment (tomasMizera) [inline app/codefilter.h:37]: Can we remove this? It seems not to be used anywhere

File: app/qml/CodeReader.qml [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+  Timer {
+    id: codeReaderTimer
+    interval: 1000
+    triggeredOnStart: false
+    repeat: false
+    onRunningChanged: {
+      if (!running) {
+        codeReader.visible = false
+      }
+    }
```
Comment (tomasMizera) [inline app/qml/CodeReader.qml:110]:
```suggestion
    onTriggered: codeReader.visible = false
```

Comment (tomasMizera) [reply app/qml/CodeReader.qml:110]: This should work too :)

**Review by tomasMizera — APPROVED** [review]
Thanks for resolving the notes, however, I unresolved one, I could not find how it was resolved (the one with `Component.onCompleted`)

Other than that there are some notes about memory. After clearing these I am ok to merge so passing to you @wonder-sk 🟢

File: app/codefilter.cpp (hunk 3 above)
Comment (sklencar) [reply app/codefilter.cpp:64]: From documentation, it seems that destructions is handled. Also added docs to the function.
http://doc.qt.io/qt-5/qabstractvideofilter.html#createFilterRunnable

File: qgsquick/from_qgis/plugin/qgsquickfeatureform.qml [hunk 7]
```diff
@@ -474,6 +497,10 @@ Item {
                return form.loadWidgetFn(widget.toLowerCase())
             else return ''
           }
+
+          Component.onCompleted: {
+            supportDataImport = importDataHandler.supportImportData(Name)
```
[the remark this answers is not in the document]
Comment (sklencar) [reply qgsquick/from_qgis/plugin/qgsquickfeatureform.qml:502]: It is simplified to use the function from handler directly to assign `supportDataImport`  attribute. Added also default function to default handler definition.

**Review by tomasMizera — APPROVED** [review]

**Pull request merged**